rfcs/2022-05-17-11532-chronicle-sink.md

```diff
@@ -0,0 +1,239 @@
+  found [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#metadata).
+  Mandatory fields are `event_type` and `event_timestamp`.
+
+  The value of `event_type` has repercussions for which fields are mandatory
+  in the rest of the event. See [here](https://cloud.google.com/chronicle/docs/unified-data-model/udm-usage#required_and_optional_fields_based_on_event_type)
+  for the list. For example, if `event_type == "FILE_COPY"` then `src.file`
+  becomes mandatory. **Currently it is not possible to validate a field based
+  on the value of another field within Vector schema, so this will need to
+  be left for future improvement.
+
+- *Principal*
+  The principal is the entity that originates the event. This field is mandatory
+  and can contain any of the fields defined
+  [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#noun).
+
+- *Src*
+  The source entity being acted upon. Depending on the `event_type` this field
+  is optional. Can contain any of the fields defined
+  [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#noun).
+
+- *Target*
+  The target entity being acted upon. Depending on the `event_type` this field
+  is optional. Can contain any of the fields defined
+  [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#noun).
+
+### Implementation
+
+Both event types require different batching strategies.
+
+#### UDM
+
+UDM can use a simple `BatchedHttpSink` with a `JsonArrayBuffer` to perform the
+batching.
+
+#### Unstructured log events
+
+Since log events are batched by `log_type` we will need to partition the
+batches on the value of the `log_type` field using a `PartitionHttpSink`.
+
+## Rationale
+
+There is significant user demand for sending data into Google Chronicle.
+
+## Drawbacks
+
+Adding a new sink does place an additional maintenance burden on the team. In
+particular since Chronicle is a third party service, developing integration
+tests to ensure the sink continues to work correctly is burdensome.
+
+## Prior Art
+
+## Alternatives
+
+An option that was considered was to use the HTTP sink and and enhance it to
+allow GCP authentication. This would nearly work for UDM events, but
+unfortunately since the payload is a JSON object with a `customer_id` field
+and an array of `events`, the structure is slightly different to what is
+possible with the HTTP sink.
+
+## Outstanding Questions
+
+Should we crate one sink with an `event_type` option, or should these be two
+separate sinks?
+
+The `log_types` field has to contain a value that has been returned from the
+`log_types` endpoint. It could be possible for the Sink to call this on boot
+time which would provide a list of possible values that could be validated
+before sending to chronicle. I feel it isn't worth doing this as the event
+will be validated by Chronicle anyway so we may as well rely on the error
+returned when sending an invalid `log_type`. Others may disagree. Is it worth
```

Comment:
Do we know if the error code/message returned in that case will be enough to exactly pinpoint a wrong `log_type` ? As events flowing here will be security events, I think we should do our best avoid preventable errored payloads.

Reply:
The errors are listed [here](https://cloud.google.com/apis/design/errors#handling_errors). I think we would probably get a 400 with an invalid log type, but will need to confirm when I get access.

Reply:
Anyway the doc seems to be lacking detailed information at the moment (there is reference to a v1 API endpoint https://cloud.google.com/chronicle/docs/supported-datasets, and there are some field like [`entity.additional`](https://cloud.google.com/chronicle/docs/reference/udm-field-list) that seems to be opaque, but somehow linked to the request `log_type` (as per [this](https://cloud.google.com/chronicle/docs/ingestion/parser-list/guardduty-changelog). I guess real life testing is probably the quickest way to get an exact answer on behaviour when an event is submitted with an invalid `log_type`.